ldap/servers/plugins/replication/replutil.c

```
@@ -668,8 +668,7 @@ repl_set_mtn_state_and_referrals(
         }
         /* We should delete referral only if we want to set the
            replica database in backend state mode */
-        /* if chain on update mode, go ahead and set the referrals anyway */
-        if (strcasecmp(mtn_state, STATE_BACKEND) == 0 || chain_on_update) {
+        if (strcasecmp(mtn_state, STATE_BACKEND) == 0) {
```

I wonder if the problem could be that the second call to repl_set_mtn_state_and_referrals (after total update) should not be with 'referrals' (provided by the total init req) instead of 'mtnreferral' (value of mapping tree referral before the total init).
In short, in multisupplier_extop_StartNSDS50ReplicationRequest, calling 
`repl_set_mtn_state_and_referrals(repl_root_sdn, mtnstate,
                                                 NULL, NULL, referrals);`

---

@tbordaz this is the stack where the referrals are NULL.  When referrals are NULL we try to delete the existing referrals which is explicitly denied by the mapping tree code:

```
#0  repl_set_mtn_state_and_referrals (repl_root_sdn=0x7fc2c78006c0, mtn_state=0x7fc808dc4e37 "referral on update", ruv=0x0, 
    ruv_referrals=0x0, other_referrals=0x0) at ../389-ds-base/ldap/servers/plugins/replication/replutil.c:608
#1  0x00007fc808d51465 in consumer5_set_mapping_tree_state_for_replica (r=0x7fc2ce05f340, supplierRuv=0x0)
    at ../389-ds-base/ldap/servers/plugins/replication/repl5_replica.c:1543
#2  0x00007fc808d51965 in replica_reload_ruv (r=0x7fc2ce05f340) at ../389-ds-base/ldap/servers/plugins/replication/repl5_replica.c:1713
#3  0x00007fc808d5708c in replica_enable_replication (r=0x7fc2ce05f340)
    at ../389-ds-base/ldap/servers/plugins/replication/repl5_replica.c:3946
#4  0x00007fc808d4bf3c in multisupplier_be_state_change (handle=0x7fc808d4bebc <multisupplier_be_state_change>, 
    be_name=0x7fc2c6604a00 "userroot", old_be_state=2, new_be_state=1)
    at ../389-ds-base/ldap/servers/plugins/replication/repl5_plugins.c:1451
#5  0x00007fc80ce747b3 in mtn_be_state_change (be_name=0x7fc2c6604a00 "userroot", old_state=2, new_state=1)
    at ../389-ds-base/ldap/servers/slapd/mapping_tree.c:218
#6  0x00007fc80ce7b54d in mtn_internal_be_set_state (be=0x7fc808867ac0, state=1) at ../389-ds-base/ldap/servers/slapd/mapping_tree.c:3610
#7  0x00007fc80ce7b5d1 in slapi_mtn_be_enable (be=0x7fc808867ac0) at ../389-ds-base/ldap/servers/slapd/mapping_tree.c:3660
#8  0x00007fc808465a28 in dbmdb_import_all_done (job=0x7fc2ca443000, ret=0)
    at ../389-ds-base/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import.c:804
#9  0x00007fc8084666be in dbmdb_public_dbmdb_import_main (arg=0x7fc2ca443000)
    at ../389-ds-base/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import.c:1133
#10 0x00007fc808466766 in dbmdb_import_main (arg=0x7fc2ca443000) at ../389-ds-base/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import.c:1169
```

Note the mapping tree code rejects the delete if the backend state  is "referral" or "referral on update"


```
mapping_tree_entry_modify_callback()

1109:
            } else if (SLAPI_IS_MOD_DELETE(mods[i]->mod_op)) {
                /* it is not OK to delete the referrals if they are still
                 * used
                 */
                if ((node->mtn_state == MTN_REFERRAL) ||
                    (node->mtn_state == MTN_REFERRAL_ON_UPDATE)) {
                    PR_snprintf(returntext, SLAPI_DSE_RETURNTEXT_SIZE,
                                "cannot delete referrals in this state\n");
                    *returncode = LDAP_UNWILLING_TO_PERFORM;
                    mtn_unlock();
                    slapi_sdn_free(&subtree);
                    slapi_ch_free_string(&plugin_fct);
                    slapi_ch_free_string(&plugin_lib);
                    return SLAPI_DSE_CALLBACK_ERROR;
                }

                mtn_free_referral_in_node(node);
```

---

Also worth noting.  We call repl_set_mtn_state_and_referrals() 3 times when doing an init:

[1]  Start repl session (replace referrals)
[2]  End repl session (replace referrals)
[3]  Enabling replication after import (this is where RUV's are null and we try to delete referrals)

---

Okay, I thought it was called twice and my suggestion is about that [line](https://github.com/389ds/389-ds-base/blob/main/ldap/servers/plugins/replication/repl_extop.c#L999) asking if changing  'mtnreferral' into referrals' could fix the issue.

---

slapi_mtn_set_referral is only called from repl_set_mtn_state_and_referrals
 and imho the modify should not fail in referral on update mode (we should be able to update the RUV entry)
  I think that operation flag should not be 0 but something else like:
    OP_FLAG_REPLICATED | OP_FLAG_REPL_FIXUP | OP_FLAG_TOMBSTONE_ENTRY | OP_FLAG_REPL_RUV as in replica_write_ruv
  
   what puzzles me is that slapi_mtn_set_referral seems to directly write the db ruv entry rather than the in memory ruv then calling replica_write_ruv
    but that may be normal (the in memory ruv may get updated from the db when enabling the backend)

---

> I think that operation flag should not be 0 but something else like:
OP_FLAG_REPLICATED | OP_FLAG_REPL_FIXUP | OP_FLAG_TOMBSTONE_ENTRY | OP_FLAG_REPL_RUV as in replica_write_ruv

This is not the issue.  We check the backend state and reject it before trying the modify operation in mapping_tree_entry_modify_callback()

---

Ok, it makes more sense. so the problem is behind the test: https://github.com/389ds/389-ds-base/blob/bf48b85bad5d82c40f1170cae1dc4a38276d161d/ldap/servers/slapd/mapping_tree.c#L1194
IMHO the idea behind test is to prevent setting REFERRAL/REFERRAL ON UPDATE mode without any referrals
but the way the test is written is too strict and reject any value deletion.
We should better accept such modifier if  slapi_entry_attr_find(entryAfter, "nsslapd-referral", &state_attr))
  exists and avoid rejecting the modify operation

---

Trying this, and returning the "|| chain_on_update" in replutil.c, causes the error to move to the supplier instead of the consumer.  So I just pushed another change and it's based solely on the comments in the code.  So we reject the delete is referrals ARE present and the backend state is a referral type.  That's what it says it's trying to do, and I have not been able to reproduce the error in any of the CI tests and my personal chain_on_update setup.

Alternatively we could also keep the code the same and just ignore error 53 in replutil to avoid the error being logged.